sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PushDownLeftSemiAntiJoin.scala

```diff
@@ -0,0 +1,203 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans._
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * This rule is a variant of [[PushDownPredicate]] which can handle
+ * pushing down Left semi and Left Anti joins below the following operators.
+ *  1) Project
+ *  2) Window
+ *  3) Union
+ *  4) Aggregate
+ *  5) Other permissible unary operators. please see [[PushDownPredicate.canPushThrough]].
+ */
+object PushDownLeftSemiAntiJoin extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    // LeftSemi/LeftAnti over Project
+    case Join(p @ Project(pList, gChild), rightOp, LeftSemiOrAnti(joinType), joinCond, hint)
+      if pList.forall(_.deterministic) && !pList.exists(ScalarSubquery.hasScalarSubquery) &&
```

Comment:
Continue the discussion.

After some more thoughts, I'm wondering why the `PushDownPredicate` doesn't care about scalar subquery at all. 

The thing I'm worried about is, currently correlated subquery filter is pushed down in the rule `PushDownPredicate` and then converted to left semi/anti joins. After your refactor, correlated subquery filter will be converted to left semi/anti join first, then be pushed down with this rule. If this rule can't behave exactly the same with `PushDownPredicate`, we may have perf regression.

Reply:
@cloud-fan I agree with your concern. Here is what i am thinking : 
1) Now the additional restriction we put in this pr is on scalar subquery.
2) If they are co-releated, then we know we will push down left semi/left anti joins below Joins in the next PR. We can see the code changes in the [link](https://github.com/apache/spark/pull/23211). So we will come up with same or better plans.
3) If they are non-corelated, then we have a concern now. This is the scenario i wanted to study and discuss and address as follow-up as i have some questions on the plan. 

Please let me know your thoughts. Another approach is to allow pushdown for non-correlated case now and fix up if necessary.

Reply:
> Another approach is to allow pushdown for non-correlated case now and fix up if necessary.

Let's go with this. If we do have a problem, we need to fix `PushDownPredicate` and this rule together.

Reply:
@cloud-fan Thank you very much. I will change.